src/neorados/cls/common.h

```diff
@@ -0,0 +1,105 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2019 Red Hat <contact@redhat.com>
+ * Author: Adam C. Emerson
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+
+#pragma once
+
+#include <boost/asio/async_result.hpp>
+#include <boost/system/error_code.hpp>
+
+#include "include/neorados/RADOS.hpp"
+
+#include "include/buffer.h"
+#include "include/encoding.h"
+
+#include "common/async/bind_like.h"
+
+#include "common/allocate_unique.h"
+
+/// \file neorados/cls/common.h
+///
+/// \brief Helpers for writing simple CLS clients
+///
+/// For writing functions that call out to the OSD, perform a CLS
+/// call, and return its data in a user friendly way, we need to
+/// coordinate asynchronous operation with decoding and returning the
+/// relevant data.
+///
+/// Ideally, these functions will try to help.
+
+namespace neorados::cls {
+/// \brief Perform a CLS read operation and return the result
+///
+/// Asynchronously call into the OSD, decoding its response and
+/// passing it to a supplied function to extract the relevant
+/// information.
+///
+/// \tparam Rep The type of the CLS operation's result that will be
+///             passed to `f`
+///
+/// \param r RADOS handle
+/// \param od Object name
+/// \param ioc IOContext locating object
+/// \param cls Name of the object class
+/// \param method Object class method to call
+/// \param f Function to extract/process call result
+/// \param token Boost.Asio CompletionToken
+///
+/// @returns The relevant data in a way appropriate to the completion
+/// token. See Boost.Asio documentation. The signature is the return
+/// type of `f`.
+template<typename Rep, typename F, typename CompletionToken>
+auto exec(
+  RADOS& r,
+  Object oid,
+  IOContext ioc,
+  std::string cls,
+  std::string method,
+  F&& f,
+  CompletionToken&& token)
+{
+  namespace ba = boost::asio;
+  namespace bs = boost::system;
+  namespace cb = ceph::buffer;
+  auto executor = ba::get_associated_executor(token);
+  return ba::async_compose<CompletionToken,
+			   void(bs::error_code, std::invoke_result_t<F, Rep>)>
+    (
+      [executor, &r, oid, ioc=std::move(ioc), cls, method,
+       f = std::move(f)](auto& self) {
+	co_spawn(
+	  executor,
+	  [self = std::move(self), &r, oid, ioc = std::move(ioc),
+	   cls = std::move(cls), method = std::move(method),
+	   f = std::move(f)]() mutable -> ba::awaitable<void> {
```

Comment:
something to be aware of when combining lambdas and coroutines:
https://devblogs.microsoft.com/oldnewthing/20211103-00/?p=105870
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rcoro-capture

Reply:
Thank you. That actually explains the bug that was confusing me.

Also, eww.

Reply:
💩